tempesta_fw/apm.c

```diff
@@ -0,0 +1,1041 @@
+			return;
+		__tfw_apm_rbent_reset(crbent, jtmistamp);
+	}
+}
+
+/*
+ * Update the control information on the APM ring buffer entries with
+ * stats for subsequent calculation of percentiles. Use optimizations
+ * to avoid the recalculation whenever possible. Maintain values of
+ * @entry_cnt, @total_cnt, @jtmwstamp that are used in optimizations.
+ *
+ * Return true if recalculation of percentiles is required.
+ * Return false if the percentile values don't need the recalculation.
+ */
+static bool
+tfw_apm_rbctl_update(TfwApmData *data, int recalc)
+{
+	int i, centry;
+	unsigned long jtmnow = jiffies;
+	unsigned long jtmwstart, jtmistart;
+	unsigned long entry_cnt, total_cnt = 0;
+	TfwApmRBuf *rbuf = &data->rbuf;
+	TfwApmRBCtl *rbctl = &data->rbctl;
+	TfwApmRBEnt *rbent = rbuf->rbent;
+
+	/* The start of the current time interval. */
+	jtmistart = jtmnow - (jtmnow % tfw_apm_jtmintrvl);
+	/* The start of the current time window. */
+	jtmwstart = jtmistart - tfw_apm_jtmwindow;
+	/* The index of the current entry. */
+	centry = (jtmnow / tfw_apm_jtmintrvl) % rbuf->rbufsz;
+
+	/*
+	 * If the latest percentiles are for a different time window,
+	 * then a recalculation is in order.
+	 */
+	if (unlikely(rbctl->jtmwstamp != jtmwstart)) {
+		tfw_apm_rbent_checkreset(&rbent[centry], jtmistart);
+
+		for (i = 0; i < rbuf->rbufsz; ++i)
+			total_cnt += atomic64_read(&rbent[i].pcntrng.total_cnt);
+
+		rbctl->entry_cnt = 0;
+		rbctl->total_cnt = total_cnt;
+		rbctl->jtmwstamp = jtmwstart;
+
+		TFW_DBG3("%s: New time window: centry [%d] total_cnt [%lu]\n",
+			 __func__, centry, rbctl->total_cnt);
+
+		return true;
+	}
+
+	/* The latest percentiles are for the current time window.
+	 * In some cases a recalculation is not required. In some
+	 * other cases the recalculation set up can be simpler.
+	 */
+
+	/* Nothing to do if there were no stats updates. */
+	entry_cnt = atomic64_read(&rbent[centry].pcntrng.total_cnt);
+	if (unlikely(rbctl->entry_cnt == entry_cnt)) {
+		if (unlikely(recalc)) {
+			TFW_DBG3("%s: Old time window: recalculate: "
+				 "centry [%d] total_cnt [%lu]\n",
+				 __func__, centry, rbctl->total_cnt);
+			TFW_INC_STAT_BH(apm.calc_repeats);
+			return true;
+		}
+		return false;
+	}
+	BUG_ON(rbctl->entry_cnt > entry_cnt);
+
+	/* Update the counts incrementally. */
+	rbctl->total_cnt += entry_cnt - rbctl->entry_cnt;
+	rbctl->entry_cnt = entry_cnt;
+
+	TFW_DBG3("%s: Old time window: centry [%d] total_cnt [%lu]\n",
+		 __func__, centry, rbctl->total_cnt);
+
+	return true;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ *
+ * Return the number of percentile values that have been filled
+ * if potentially new percentile values were calculated.
+ * Return 0 if the percentile values didn't need the recalculation.
+ * REturn -1 of the recalculation could not be performed.
+ */
+static int
+__tfw_apm_calc(TfwApmData *data, TfwApmPrcntl *prcntl, int recalc)
+{
+	int ret;
+
+	if (!spin_trylock(&tfw_sa_guard)) {
```

Comment:
This lock is needed to prevent range changes while the percentiles are calculated. The range changes may occur only for the current ring buffer entry, and that ring buffer entry is the same for all servers. I don't see any practical reason for having a lock per server data structure.